subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinDslAssignmentIntegrationTest.kt

```diff
@@ -181,6 +181,63 @@ class KotlinDslAssignmentIntegrationTest : AbstractKotlinIntegrationTest() {
         build("myTask")
     }
 
+    @Test
+    fun `assign operator compiles with all possible Property types`() {
+        // Given
+        withAssignmentOverload()
+        withBuildScript("""
+            abstract class MyTask : DefaultTask() {
+                @get:Input
+                abstract val input: Property<String>
+                @get:Input
+                abstract val mapInput: MapProperty<String, String>
+                @get:Input
+                abstract val setInput: SetProperty<String>
+                @get:Input
+                abstract val listInput: ListProperty<String>
+                @get:InputDirectory
+                abstract val dirInput: DirectoryProperty
+                @get:OutputFile
+                abstract val fileOutput: RegularFileProperty
+
+                @TaskAction
+                fun taskAction() {
+                    fileOutput.asFile.get().writeText(input.get())
+                }
+            }
+
+            tasks.register<MyTask>("myTask") {
+                file("src").mkdirs()
+                input = null
+                input = "Hello world"
+                input = provider { null }
+                input = provider { "Hello" }
+                mapInput = null
+                mapInput = mapOf("a" to "b")
+                mapInput = provider { null }
+                mapInput = provider { mapOf("a" to "b") }
+                setInput = null
+                setInput = listOf("a")
+                setInput = provider { null }
+                setInput = provider { listOf("a") }
+                listInput = null
+                listInput = listOf("a")
+                listInput = provider { null }
+                listInput = provider { listOf("a") }
+                dirInput = null as File?
```

Comment:
> I assume dirInput = null fails because the compiler can't disambiguate.

Yes exactly. 
Because we have:
```
fun <T> Property<T>.assign(value: T?)

and

fun <T : FileSystemLocation> FileSystemLocationProperty<T>.assign(file: File?)
```

I need to check for groovy.

----

FYI for Provider I implemented:
```
fun <T> Property<T>.assign(value: Provider<out T?>)
```

So if we want to support:
```
dirInput = null as Provider<File>?
```

I need to change it to:
```
fun <T> Property<T>.assign(value: Provider<out T?>?)
```

Do we want to allow null provider?

Reply:
I guess we should reflect how `.set()` works so we should also have:
```
fun <T> Property<T>.assign(value: Provider<out T?>?)
```

Reply:
Yes, matching the Java API sounds reasonable :+1:

Reply:
Note: when I use:
```
fun <T> Property<T>.assign(value: Provider<out T?>?)
```

I get a warning:
<img width="501" alt="Screenshot 2023-01-18 at 18 14 38" src="https://user-images.githubusercontent.com/3939893/213252593-03ab4743-0e24-4036-a89b-5859cb08a9d4.png">

So it seems compiler can see our Java definition as non-null.